llama_stack/apis/memory_banks/memory_banks.py

```diff
@@ -20,59 +22,121 @@ class MemoryBankType(Enum):
+class KeywordRegistration(BaseRegistration):
+    pass
 
 
 @json_schema_type
-class GraphMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.graph.value] = MemoryBankType.graph.value
-
-
-MemoryBankDef = Annotated[
-    Union[
-        VectorMemoryBankDef,
-        KeyValueMemoryBankDef,
-        KeywordMemoryBankDef,
-        GraphMemoryBankDef,
-    ],
-    Field(discriminator="type"),
+class GraphRegistration(BaseRegistration):
+    pass
+
+
+RegistrationRequest = Union[
+    VectorRegistration,
+    KeyValueRegistration,
+    KeywordRegistration,
+    GraphRegistration,
 ]
 
-MemoryBankDefWithProvider = MemoryBankDef
+
+def registration_request_to_memory_bank(request: RegistrationRequest) -> MemoryBank:
+    """Convert registration request to memory bank object"""
+    if isinstance(request, VectorRegistration):
+        return VectorMemoryBank(
+            identifier=request.memory_bank_id,
+            provider_resource_id=request.provider_resource_id,
+            provider_id=request.provider_id,
+            embedding_model=request.embedding_model,
+            chunk_size_in_tokens=request.chunk_size_in_tokens,
+            overlap_size_in_tokens=request.overlap_size_in_tokens,
+        )
+    elif isinstance(request, KeyValueRegistration):
+        return KeyValueMemoryBank(
+            identifier=request.memory_bank_id,
+            provider_resource_id=request.provider_resource_id,
+            provider_id=request.provider_id,
+            memory_bank_type=MemoryBankType.keyvalue,
+        )
+    elif isinstance(request, KeywordRegistration):
+        return KeywordMemoryBank(
+            identifier=request.memory_bank_id,
+            provider_resource_id=request.provider_resource_id,
+            provider_id=request.provider_id,
+            memory_bank_type=MemoryBankType.keyword,
+        )
+    elif isinstance(request, GraphRegistration):
+        return GraphMemoryBank(
+            identifier=request.memory_bank_id,
+            provider_resource_id=request.provider_resource_id,
+            provider_id=request.provider_id,
+            memory_bank_type=MemoryBankType.graph,
+        )
+    else:
+        raise ValueError(f"Unknown registration type: {type(request)}")
 
 
 @runtime_checkable
 class MemoryBanks(Protocol):
     @webmethod(route="/memory_banks/list", method="GET")
-    async def list_memory_banks(self) -> List[MemoryBankDefWithProvider]: ...
+    async def list_memory_banks(self) -> List[MemoryBank]: ...
 
     @webmethod(route="/memory_banks/get", method="GET")
-    async def get_memory_bank(
-        self, identifier: str
-    ) -> Optional[MemoryBankDefWithProvider]: ...
+    async def get_memory_bank(self, memory_bank_id: str) -> Optional[MemoryBank]: ...
 
     @webmethod(route="/memory_banks/register", method="POST")
     async def register_memory_bank(
-        self, memory_bank: MemoryBankDefWithProvider
-    ) -> None: ...
+        self, request: RegistrationRequest
```

Comment:
hm, why is this API not shaped like the other ones like model / shield?

Reply:
because we have different subtypes of memory banks each with different params.

Reply:
I think we still should not create this new Registration type. We should have an Annotated union for the bank and destructure the common parameters out into the API call. Here's what it should look like:

```
# better name than a BankDef
BankDef = Annotated[
  Union[ SpecificBank1, SpecificBank2 ],
  Field(discriminator="type")
]

register_memory_bank(
  self,
  memorybank_id,
  provider_id,
  provider_memorybank_id,
  bank: Bank 
)
```

basically we should not have this term "Request" show up in the API. It makes this API look very different than others without a strong reason.

Reply:
done.